logic/Rule.java

```diff
@@ -472,30 +472,91 @@ public Relation(RelationConstraint relation, IsaConstraint isa, Rule rule) {
-                        return thenConcepts;
+                FunctionalIterator<com.vaticle.typedb.core.concept.thing.Relation> existingRelations =
+                        matchRelation(relationType, whenConcepts, traversalEng, conceptMgr);
+                if (!existingRelations.hasNext()) {
+                    return Optional.of(insert(relationType, whenConcepts));
+                } else {
+                    while (existingRelations.hasNext()) {
+                        com.vaticle.typedb.core.concept.thing.Relation preexisting = existingRelations.next();
+                        if (!preexisting.isInferred()) return Optional.empty();
+                        else {
+                            if (insertable(preexisting, whenConcepts)) {
+                                return Optional.of(thenConcepts(preexisting, whenConcepts));
+                            }
+                        }
+                    }
+                    return Optional.empty();
+                }
+            }
+
+            private boolean insertable(
+                    com.vaticle.typedb.core.concept.thing.Relation inserted, ConceptMap whenConcepts) {
+                if (isa.type().label().isPresent()) {
+                    if (!inserted.getType().getLabel().equals(isa.type().label().get().properLabel())) return false;
+                } else if (!inserted.getType().getLabel().equals(whenConcepts.get(isa.type().id().asRetrievable()).asType().getLabel())) return false;
+                Map<String, Map<Concept, Integer>> relationMap = new HashMap<>();
+                relation.players().forEach(player -> {
+                    assert player.roleType().isPresent(); // Must be present to be insertable
+                    TypeVariable roleType = player.roleType().get();
+                    String descopedRoleTypeLabel = roleType.label().map(LabelConstraint::properLabel)
+                            .orElseGet(() -> whenConcepts.get(roleType.id().asRetrievable()).asType().getLabel()).name();
+                    Map<Concept, Integer> played = relationMap.computeIfAbsent(descopedRoleTypeLabel, p -> new HashMap<>());
+                    Concept playerConcept = whenConcepts.get(player.player().id());
+                    played.computeIfPresent(playerConcept, (p, count) -> count + 1);
+                    played.putIfAbsent(playerConcept, 1);
+                });
+                Map<String, Map<Concept, Integer>> insertedMap = new HashMap<>();
+                inserted.getPlayersByRoleType().forEach((role, players) -> {
+                    players.forEach(player -> {
+                        Map<Concept, Integer> playerEntry = insertedMap.computeIfAbsent(role.getLabel().name(), r -> new HashMap<>());
+                        playerEntry.computeIfPresent(player, (p, count) -> count + 1);
+                        playerEntry.putIfAbsent(player, 1);
                     });
+                });
+                return relationMap.equals(insertedMap);
```

Comment:
To do the comparison in an easy-to-understand way, we just create a map for each and compare them. Seemed less error prone than trying to iterate over one and compare with the other while making sure both have the same counts.

Reply:
I've been able to get to the 16 lines you describe, but only without incorrect equality. Equality for a map of lists isn't what we want here. Really, the smell is that `getPlayersByRoleType` gives a map of lists rather than a map of multisets imo, then working with it and doing the comparison would be easy.